miden/src/repl/mod.rs

```diff
@@ -0,0 +1,317 @@
+                                // each memory register converted from Felt to an integer.
+                                let mem_int =
+                                    memory_value.iter().map(|&x| x.as_int()).collect::<Vec<_>>();
+                                // displays the state of the memory at index x if present.
+                                println!("{}: {:?}", memory_at_addr, mem_int);
+                                flag = true;
+                                break;
+                            }
+                        }
+                        if !flag {
+                            println!("Memory at address {} is empty", memory_at_addr,);
+                        }
+                    }
+                    stack_flag = false;
+                } else if line == "!undo" {
+                    let last_line = program_lines.pop().unwrap();
+                    println!("Undoing {}", last_line);
+                    stack_flag = true;
+                } else if line == "!stack" {
+                    println!("{}", result_string);
+                    stack_flag = false;
+                } else {
+                    rl.add_history_entry(line.clone());
+                    program_lines.push(line.clone());
+                    stack_flag = true;
+                }
+            }
+            Err(ReadlineError::Interrupted) => {
+                println!("CTRL-C");
+                break;
+            }
+            Err(ReadlineError::Eof) => {
+                println!("CTRL-D");
+                break;
+            }
+            Err(err) => {
+                println!("Error: {:?}", err);
+                break;
+            }
+        };
+    }
+    rl.save_history("history.txt").unwrap();
+}
+
+/// HELPER METHODS
+/// --------------------------------------------------------------------------------------------
+
+/// Compiles and executes a compiled Miden program, returning the stack, memory and any Miden errors.
+/// The program is passed in as a String, passed to the Miden Assembler, and then passed into the Miden
+/// Processor to be executed.
+pub fn execute(
+    program: String,
+    _pub_inputs: Vec<u128>,
+) -> Result<(Vec<(u64, Word)>, String), MidenError> {
+    let program = assembly::Assembler::new(true)
+        .compile(&program)
+        .map_err(MidenError::AssemblyError)?;
+
+    let pub_inputs = vec![];
+    let inputs = ProgramInputs::new(&pub_inputs, &[], vec![]).unwrap();
+    let mut process = Process::new_debug(inputs);
+    let _program_outputs = process
+        .execute(&program)
+        .map_err(MidenError::ExecutionError);
+
+    let (sys, _, stack, _, chiplets) = process.to_components();
+    let mem = chiplets.get_mem_state_at(0, sys.clk());
+
+    let stack_state = format!(
+        "{}",
+        stack
+            .get_state_at(sys.clk())
+            .iter()
+            .map(|f| format!("{}", f))
+            .collect::<Vec<_>>()
+            .join(" "),
+    );
```

Comment:
I would move this into a separate helper function - something like `print_stack()`. And then we could change the signature of this function to be like:
```Rust
fn execute(program: String) -> Result<(Vec<(u64, Word)>, Vec<Felt>), MidenError>
```